File: ckgame-java/src/main/java/com/ckgame/politics/Faction.java
package com.ckgame.politics;

import com.ckgame.core.Constants;

import java.util.ArrayList;
import java.util.List;

/**
 * 派系对象。
 */
public final class Faction {
    public final int id;
    public final FactionKind kind;
    public final int targetLiege;
    public final List<Integer> members = new ArrayList<>();
    public double power = 0.0;
    public double discontent = 0.0;
    public boolean ultimatumSent = false;
    public int claimant = Constants.NONE_ID;

    public Faction(int id, FactionKind kind, int targetLiege, Integer claimant) {
        this.id = id;
        this.kind = kind;
        this.targetLiege = targetLiege;
        if (claimant != null) {
            this.claimant = claimant;
        }
    }

    /** 添加成员（去重）。 */
    public void addMember(int who) {
        if (!members.contains(who)) {
            members.add(who);
        }
    }

    /** 移除成员。 */
    public void removeMember(int who) {
        members.remove(Integer.valueOf(who));
    }

    /** 是否已准备好叛乱。 */
    public boolean isReadyToRevolt() {
        return power >= 80 && discontent >= 50 && members.size() >= 2;
    }

    /** 是否可发出最后通牒。 */
    public boolean canSendUltimatum() {
        return power >= 60
                && discontent >= 40
                && !ultimatumSent
                && members.size() >= 2;
    }
}

File: ckgame-ts/src/politics/Faction.ts
import { NONE_ID } from '../core/Constants.js';
import { FactionKind } from './FactionKind.js';

/**
 * A faction object.
 */
export class Faction {
  public readonly id: number;
  public readonly kind: FactionKind;
  public readonly targetLiege: number;
  public members: number[] = [];
  public power: number = 0;
  public discontent: number = 0;
  public ultimatumSent: boolean = false;
  public claimant: number = NONE_ID;

  constructor(id: number, kind: FactionKind, targetLiege: number, claimant?: number) {
    this.id = id;
    this.kind = kind;
    this.targetLiege = targetLiege;
    if (claimant !== undefined && claimant !== null) {
      this.claimant = claimant;
    }
  }

  /** Add a member (deduplicated). */
  addMember(who: number): void {
    if (!this.members.includes(who)) {
      this.members.push(who);
    }
  }

  /** Remove a member. */
  removeMember(who: number): void {
    const idx = this.members.indexOf(who);
    if (idx >= 0) this.members.splice(idx, 1);
  }

  /** Whether the faction is ready to revolt. */
  isReadyToRevolt(): boolean {
    return this.power >= 80 && this.discontent >= 50 && this.members.length >= 2;
  }

  /** Whether the faction can send an ultimatum. */
  canSendUltimatum(): boolean {
    return this.power >= 60
      && this.discontent >= 40
      && !this.ultimatumSent
      && this.members.length >= 2;
  }
}
